### ISSUE TITLE
String interpolation parsing in initializers regression in 3.2.0

### ISSUE THREAD

**Issue by jrudolph** [lampepfl/dotty#16043]
## Compiler version

3.2.0

## Minimized code

```scala
val str = "test"
class Test { s"$str" }
```

## Output

```
-- Error: ----------------------------------------------------------------------
2 |class Test { s"$str" }
  |                   ^
  |                   unclosed string literal
```

## Expectation

Compiles as in 3.1.x.

This seems to be a particular issue about initializers. Here are a few instances that work:

```scala
{ val str = "test"; s"$str" }
class Test(str: String) { def test = s"$str" }
{ val str = "test"; trait Test { s"${str}" } }
```

**Comment by jrudolph** [comment]
Seems to be fixed in 3.2.1-RC1 so just filing for reference. Maybe same/similar cause as #15514?

**Comment by som-snytt** [comment]
Yes, that is the same issue. The shared element is the string interpolation in "self-type" position of a template.
